In [1]:
import os
dataset_path = 'cicids2017_cleaned.csv'

# Task
Train a Random Forest model on the CICIDS2017 dataset located at "Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv" and evaluate its performance.

## Load data

### Subtask:
Load the CICIDS2017 dataset from the specified path into a pandas DataFrame.


**Reasoning**:
Import pandas and load the dataset into a DataFrame.



In [2]:
import pandas as pd

df = pd.read_csv(dataset_path)
df.head()

,Destination Port,Flow Duration,Total Fwd Packets,Total Length of Fwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,Bwd Packet Length Max,Bwd Packet Length Min,...,Init_Win_bytes_backward,act_data_pkt_fwd,min_seg_size_forward,Active Mean,Active Max,Active Min,Idle Mean,Idle Max,Idle Min,Attack Type
0,22,1266342,41,2664,456,0,64.975610,109.864573,976,0,...,243,24,32,0.0,0,0,0.0,0,0,Normal Traffic
1,22,1319353,41,2664,456,0,64.975610,109.864573,976,0,...,243,24,32,0.0,0,0,0.0,0,0,Normal Traffic
2,22,160,1,0,0,0,0.000000,0.000000,0,0,...,243,0,32,0.0,0,0,0.0,0,0,Normal Traffic
3,22,1303488,41,2728,456,0,66.536585,110.129945,976,0,...,243,24,32,0.0,0,0,0.0,0,0,Normal Traffic
4,35396,77,1,0,0,0,0.000000,0.000000,0,0,...,290,0,32,0.0,0,0,0.0,0,0,Normal Traffic


In [3]:
print(df.columns.tolist())
print(len(df.columns))

['Destination Port', 'Flow Duration', 'Total Fwd Packets', 'Total Length of Fwd Packets', 'Fwd Packet Length Max', 'Fwd Packet Length Min', 'Fwd Packet Length Mean', 'Fwd Packet Length Std', 'Bwd Packet Length Max', 'Bwd Packet Length Min', 'Bwd Packet Length Mean', 'Bwd Packet Length Std', 'Flow Bytes/s', 'Flow Packets/s', 'Flow IAT Mean', 'Flow IAT Std', 'Flow IAT Max', 'Flow IAT Min', 'Fwd IAT Total', 'Fwd IAT Mean', 'Fwd IAT Std', 'Fwd IAT Max', 'Fwd IAT Min', 'Bwd IAT Total', 'Bwd IAT Mean', 'Bwd IAT Std', 'Bwd IAT Max', 'Bwd IAT Min', 'Fwd Header Length', 'Bwd Header Length', 'Fwd Packets/s', 'Bwd Packets/s', 'Min Packet Length', 'Max Packet Length', 'Packet Length Mean', 'Packet Length Std', 'Packet Length Variance', 'FIN Flag Count', 'PSH Flag Count', 'ACK Flag Count', 'Average Packet Size', 'Subflow Fwd Bytes', 'Init_Win_bytes_forward', 'Init_Win_bytes_backward', 'act_data_pkt_fwd', 'min_seg_size_forward', 'Active Mean', 'Active Max', 'Active Min', 'Idle Mean', 'Idle Max', 'Id

## Prepare data

### Subtask:
Separate the features (X) and the target variable (y).


**Reasoning**:
Separate the features (X) and the target variable (y) as instructed by dropping the 'Attack Type' column for X and selecting the 'Attack Type' column for y.



In [4]:
X = df.drop('Attack Type', axis=1)
y = df['Attack Type']

display(X.head())
display(y.head())

,Destination Port,Flow Duration,Total Fwd Packets,Total Length of Fwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,Bwd Packet Length Max,Bwd Packet Length Min,...,Init_Win_bytes_forward,Init_Win_bytes_backward,act_data_pkt_fwd,min_seg_size_forward,Active Mean,Active Max,Active Min,Idle Mean,Idle Max,Idle Min
0,22,1266342,41,2664,456,0,64.975610,109.864573,976,0,...,29200,243,24,32,0.0,0,0,0.0,0,0
1,22,1319353,41,2664,456,0,64.975610,109.864573,976,0,...,29200,243,24,32,0.0,0,0,0.0,0,0
2,22,160,1,0,0,0,0.000000,0.000000,0,0,...,290,243,0,32,0.0,0,0,0.0,0,0
3,22,1303488,41,2728,456,0,66.536585,110.129945,976,0,...,29200,243,24,32,0.0,0,0,0.0,0,0
4,35396,77,1,0,0,0,0.000000,0.000000,0,0,...,243,290,0,32,0.0,0,0,0.0,0,0


0    Normal Traffic
1    Normal Traffic
2    Normal Traffic
3    Normal Traffic
4    Normal Traffic
Name: Attack Type, dtype: str

## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the data into training and testing sets and print their shapes.



In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (2016600, 52)
Shape of X_test: (504151, 52)
Shape of y_train: (2016600,)
Shape of y_test: (504151,)


## Train random forest model

### Subtask:
Train a Random Forest classifier on the training data.


**Reasoning**:
Train a RandomForestClassifier model on the training data.



In [6]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

KeyboardInterrupt: 

## Evaluate model

### Subtask:
Evaluate the performance of the trained model on the testing data.


**Reasoning**:
Evaluate the performance of the trained Random Forest model on the test data using accuracy, classification report, and confusion matrix.



In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

y_pred = rf_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.9983953220364533

Classification Report:
                precision    recall  f1-score   support

          Bots       0.90      0.76      0.82       384
   Brute Force       1.00      1.00      1.00      1821
          DDoS       1.00      1.00      1.00     25675
           DoS       1.00      1.00      1.00     38912
Normal Traffic       1.00      1.00      1.00    418697
 Port Scanning       0.99      0.99      0.99     18219
   Web Attacks       0.99      0.96      0.98       443

      accuracy                           1.00    504151
     macro avg       0.98      0.96      0.97    504151
  weighted avg       1.00      1.00      1.00    504151


Confusion Matrix:
[[   292      0      0      0     92      0      0]
 [     0   1817      0      1      3      0      0]
 [     0      0  25664      0     11      0      0]
 [     0      0      0  38784    125      1      2]
 [    34      0      1     62 418390    210      0]
 [     0      0      0      6    243  17969      

## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded, containing 52 features and a target variable 'Attack Type'.
*   The data was split into training (80%) and testing (20%) sets, resulting in 2,016,600 samples for training and 504,151 samples for testing.
*   A Random Forest classifier with 100 estimators was trained on the training data.
*   The trained Random Forest model achieved a high accuracy of approximately 0.9984 on the test set.
*   The classification report showed strong performance across most classes, with high precision, recall, and f1-scores, indicating the model is effective in distinguishing between different traffic types.

### Insights or Next Steps

*   Given the high performance, the trained model is ready for deployment to detect different types of network traffic, including attacks.
*   Further analysis could involve exploring feature importance to identify the most influential features in detecting network attacks.


## Save Model and Feature Names

### Subtask:
Save the trained Random Forest model and the list of feature column names to files.

**Reasoning**:
Save the trained Random Forest model to a `.pkl` file and the list of feature column names to a `.joblib` file using the `joblib` library.

In [ ]:
import joblib
import os

# Define the directory to save the files
save_dir = './trained_model'
os.makedirs(save_dir, exist_ok=True)

# Save the model
model_filename = os.path.join(save_dir, 'random_forest_model.pkl')
joblib.dump(rf_model, model_filename)

# Save the feature names
features_filename = os.path.join(save_dir, 'feature_names.joblib')
joblib.dump(X.columns.tolist(), features_filename)

print(f"Model saved to: {model_filename}")
print(f"Feature names saved to: {features_filename}")

Model saved to: /content/drive/MyDrive/trained_model/random_forest_model.pkl
Feature names saved to: /content/drive/MyDrive/trained_model/feature_names.joblib


## Download Files

### Subtask:
Provide code to download the saved model and feature names files.

**Reasoning**:
Use `google.colab.files.download` to download the saved model and feature names files to your local machine.

In [ ]:
from google.colab import files

# Download the model file
files.download(model_filename)

# Download the feature names file
files.download(features_filename)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>